src/details/ArborX_NeighborList.hpp

```diff
@@ -0,0 +1,159 @@
+/****************************************************************************
+ * Copyright (c) 2017-2023 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_NEIGHBOR_LIST_HPP
+#define ARBORX_NEIGHBOR_LIST_HPP
+
+#include <ArborX_DetailsHalfTraversal.hpp>
+#include <ArborX_DetailsKokkosExtViewHelpers.hpp> // reallocWithoutInitializing
+#include <ArborX_DetailsUtils.hpp>                // exclusivePrefixSum
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Sphere.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX::Experimental
+{
+
+struct NeighborListPredicateGetter
+{
+  float _radius;
+
+  KOKKOS_FUNCTION auto operator()(Box b) const
+  {
+    return intersects(Sphere{b.minCorner(), _radius});
+  }
+};
+
+template <class ExecutionSpace, class Primitives, class Offsets, class Indices>
+void findHalfNeighborList(ExecutionSpace const &space,
+                          Primitives const &primitives, float radius,
+                          Offsets &offsets, Indices &indices)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Experimental::HalfNeighborList");
+
+  using Details::HalfTraversal;
+
+  using MemorySpace =
+      typename AccessTraits<Primitives, PrimitivesTag>::memory_space;
+  BVH<MemorySpace> bvh(space, primitives);
+  int const n = bvh.size();
+
+  Kokkos::Profiling::pushRegion(
+      "ArborX::Experimental::HalfNeighborList::Count");
+
+  Kokkos::View<int *, MemorySpace> counts(
+      Kokkos::view_alloc(space,
+                         "ArborX::Experimental::HalfNeighborList::counts"),
+      n);
+  HalfTraversal(
+      space, bvh,
+      KOKKOS_LAMBDA(int, int j) { Kokkos::atomic_increment(&counts(j)); },
+      NeighborListPredicateGetter{radius});
+
+  KokkosExt::reallocWithoutInitializing(space, offsets, n + 1);
+  Kokkos::deep_copy(space, Kokkos::subview(offsets, std::make_pair(0, n)),
+                    counts);
+  exclusivePrefixSum(space, offsets);
+  KokkosExt::reallocWithoutInitializing(space, indices,
+                                        KokkosExt::lastElement(space, offsets));
+
+  Kokkos::Profiling::popRegion();
+  Kokkos::Profiling::pushRegion("ArborX::Experimental::HalfNeighborList::Fill");
+
+  Kokkos::deep_copy(space, counts, 0);
```

Comment:
Can/should use the trick pointed out in https://github.com/arborx/ArborX/pull/811#pullrequestreview-1237744400